packages/core/src/common/messaging/proxy-factory.ts

```diff
@@ -212,46 +237,51 @@ export class JsonRpcProxyFactory<T extends object> implements ProxyHandler<T> {
      *
      * @param target - unused.
-     * @param p - The property accessed on the Proxy object.
+     * @param property - The property accessed on the Proxy object.
      * @param receiver - unused.
      * @returns A callable that executes the JSON-RPC call.
      */
-    get(target: T, p: PropertyKey, receiver: any): any {
-        if (p === 'setClient') {
+    get(target: T, property: PropertyKey, receiver: any): any {
+        if (property === 'setClient') {
             return (client: any) => {
                 this.target = client;
             };
         }
-        if (p === 'getClient') {
+        if (property === 'getClient') {
             return () => this.target;
         }
-        if (p === 'onDidOpenConnection') {
+        if (property === 'onDidOpenConnection') {
             return this.onDidOpenConnectionEmitter.event;
         }
-        if (p === 'onDidCloseConnection') {
+        if (property === 'onDidCloseConnection') {
             return this.onDidCloseConnectionEmitter.event;
         }
-        const isNotify = this.isNotification(p);
-        return (...args: any[]) => {
-            const method = p.toString();
+        const isNotify = this.isNotification(property);
+        return async (...args: any[]) => {
+            const method = property.toString();
             const capturedError = new Error(`Request '${method}' failed`);
-            return this.connectionPromise.then(connection =>
-                new Promise((resolve, reject) => {
-                    try {
-                        if (isNotify) {
-                            connection.sendNotification(method, ...args);
-                            resolve();
-                        } else {
-                            const resultPromise = connection.sendRequest(method, ...args) as Promise<any>;
-                            resultPromise
-                                .catch((err: any) => reject(this.deserializeError(capturedError, err)))
-                                .then((result: any) => resolve(result));
-                        }
-                    } catch (err) {
-                        reject(err);
-                    }
-                })
-            );
+            const connection = await this.connectionPromise;
+            // Extract the cancellation token if it is the last argument passed to the method to call over JSON-RPC.
+            let token: CancellationToken | undefined;
+            if (CancellationToken.is(args[args.length - 1])) {
+                token = args.pop();
+            }
+            // Convert the list of arguments into a structure that allows us to conserve undefined values.
+            const params: any[] = this.createJsonRpcParameters(args);
+            // Push back the token if there was one.
+            // `vscode-jsonrpc` will remove the `CancellationToken` out of `params` when is is the last element.
+            if (token) {
+                params.push(token);
+            }
```

Comment:
Like I mentioned, `undefined` values are transformed into `null` without the type system knowing.

https://github.com/eclipse-theia/theia/blob/8bf15dc1d9b186b11c1f954b4c51500e611d3072/packages/plugin-ext/src/main/common/plugin-paths-protocol.ts#L24-L25

Here `workspaceUri` is actually `string | null` when used over RPC. We got "lucky" here as the only implementation checks for truthiness to detect the type rather than some `workspaceUri === undefined` which would not work.

One solution would be for interfaces meant to be used over RPC to almost never use `undefined` in their API.